### PR TITLE
Fix jira_user table to retry rate limit exceeded errors closes #68

### LINKED ISSUE #68: user_table not returning all results
**Describe the bug**

`select * from jira_user` doesn't display all records and returns an inconsistent number of records every time it is run (even without clearing the cache)

**Additional information**

- There is no error so you'd think everything worked properly
- `select count(*) from jira.jira_user` does display the correct number of users
- Adding `order by` causes a 429 error every time. Example:
`select * from jira.jira_user order by display_name`
`Error: table 'jira_user' column 'group_names' requires hydrate data from getUserGroups, which failed with error 429 Too Many Requests: local_rate_limited: request failed. Please analyze the request body for more details. Status code: 429.`
- Reducing `MaxConcurrency` seems to fix the issue. I reduced it from 50 to 5 and consistently get all records.
- Running against a Jira cloud instance with a few hundred users

**Steampipe version (`steampipe -v`)**
Example: v0.16.4

**Plugin version (`steampipe plugin list`)**
Example: v0.6.1

### PR DESCRIPTION
**Pull request by misraved**
# Example query results
<details>
  <summary>Results</summary>
  
```
Add example SQL query results here (please include the input queries as well)
```
</details>

### CODE REVIEW

File: jira/table_jira_user.go [hunk 1]
```diff
@@ -155,12 +156,13 @@ func getUserGroups(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateDa
...
 
 	groups, _, err := client.User.GetGroups(user.AccountID)
+
 	if err != nil {
 		plugin.Logger(ctx).Error("jira_user.getUserGroups", "api_error", err)
 		return nil, err
 	}
 
-	return groups, nil
+	return  groups, nil
```
Comment (cbruno10) [inline jira/table_jira_user.go:165]: Should there be an extra space here?

File: jira/table_jira_user.go [hunk 2]
```diff
@@ -26,6 +26,7 @@ func tableUser(_ context.Context) *plugin.Table {
 				// Limit concurrency to avoid a 429 too many requests error
 				Func:           getUserGroups,
 				MaxConcurrency: 50,
+				RetryConfig: 		&plugin.RetryConfig{ ShouldRetryErrorFunc: shouldRetryError([]string{"429"})},
```
Comment (cbruno10) [inline jira/table_jira_user.go:29]: Spacing looks off here.

Also, if this works for this table, should we add a `RetryConfig` for all tables?

File: jira/errors.go [hunk 3]
```diff
@@ -9,3 +15,14 @@ func isNotFoundError(err error) bool {
 func isBadRequestError(err error) bool {
 	return strings.Contains(err.Error(), "400")
 }
+
+func shouldRetryError(retryErrors []string) plugin.ErrorPredicateWithContext {
+	return func(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData, err error) bool {
+		
+		if strings.Contains(err.Error(), "429") {
+			log.Printf("[WARN] Received Rate Limit Error")
```
Comment (cbruno10) [inline jira/errors.go:23]: Is this log message actually printed out? I think it'd be better if we used the standard `plugin.Logger(ctx).LEVEL` function with the proper log message format, e.g., `functionName, message, data`.

Also, please check what we do in other plugins as reference.

**Review by cbruno10 — CHANGES_REQUESTED** [review]
@misraved I know this is still in draft, so some of my comments may be irrelevant already, but I left a few preliminary ones.

File: jira/errors.go [hunk 4]
```diff
@@ -9,3 +14,14 @@ func isNotFoundError(err error) bool {
 func isBadRequestError(err error) bool {
 	return strings.Contains(err.Error(), "400")
 }
+
+func shouldRetryError(retryErrors []string) plugin.ErrorPredicateWithContext {
+	return func(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData, err error) bool {
+
+		if strings.Contains(err.Error(), "429") {
+			plugin.Logger(ctx).Warn("Received Rate Limit Error")
```
Comment (cbruno10) [inline jira/errors.go:22]: This should be `Debug` level, not `Warn`. Also, we should use the standard log message format, e.g., https://github.com/turbot/steampipe-plugin-github/blob/main/github/errors.go#L18. I'm not sure what additional info we get (if any), but at the very least, we should include plugin/function name, and other standard args.

https://github.com/turbot/steampipe-plugin-jira/pull/72?notification_referrer_id=NT_kwDOALym4LM0ODg3MTE5ODgwOjEyMzYzNDg4&notifications_after=Y3Vyc29yOjI1#discussion_r1028166338

File: jira/plugin.go [hunk 5]
```diff
@@ -12,8 +12,9 @@ const pluginName = "steampipe-plugin-jira"
 // Plugin creates this (jira) plugin
 func Plugin(ctx context.Context) *plugin.Plugin {
 	p := &plugin.Plugin{
-		Name:             pluginName,
-		DefaultTransform: transform.FromCamel(),
+		Name:               pluginName,
+		DefaultTransform:   transform.FromCamel(),
+		DefaultRetryConfig: &plugin.RetryConfig{ShouldRetryErrorFunc: shouldRetryError([]string{"429"})},
```
Comment (cbruno10) [inline jira/plugin.go:17]: Was this question ever commented/resolved? https://github.com/turbot/steampipe-plugin-jira/pull/72#discussion_r1028165413

**Review by cbruno10 — CHANGES_REQUESTED** [review]
@misraved Please see comments, thanks!

File: jira/errors.go (hunk 4 above)
Comment (misraved) [reply jira/errors.go:22]: Sorry @cbruno10, had missed this reference to the GitHub plugin. I was following the error message format used in https://github.com/turbot/steampipe-plugin-steampipecloud/blob/main/steampipecloud/errors.go#L10

I have resolved it now 👍 .

File: jira/plugin.go (hunk 5 above)
Comment (misraved) [reply jira/plugin.go]: @cbruno10 I have fixed the formatting in dd6405fbfcb5ace6d3b7caba090b6816ae21989a

I have now added the `shouldRetryError` function to the `DefaultRetryConfig` which applies it across all the tables.

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**